woocommerce-services.php

```diff
@@ -1025,6 +1026,77 @@ public function reports_tabs( $reports ) {
 			return $reports;
 		}
 
+		/**
+		 * Send completed order details to Connect Server to check live rates usage.
+		 * Attached to the woocommerce_order_details_after_order_table hook
+		 *
+		 * @param int   $order_id int for completed order ID.
+		 * @param array $posted_data Post data for order.
+		 * @param array $order WC_Order object containing completed order details.
+		 */
+		public function track_completed_order( $order_id, $data, $order ) {
+			$logger = $this->get_logger();
+
+			$packages = WC()->cart->get_shipping_packages();
+			$package  = $packages[0];
+			foreach ( $package['contents'] as $item_key => $item_details ) {
+				ksort( $package['contents'][ $item_key ] );
+			}
+			$package['rates'] = array();
+
+			$shipping_zone    = WC_Shipping_Zones::get_zone_matching_package( $package );
+			$shipping_methods = $shipping_zone->get_shipping_methods( true );
+
+			$tracking_request_body = array();
+			foreach ( $shipping_methods as $method ) {
+				try {
+					$handler = new WC_Connect_Shipping_Method( $method->get_instance_id() );
+
+					$service_settings     = $handler->get_service_settings();
+					$service_schema       = $handler->get_service_schema();
+					$cached_response_body = $handler->get_cached_shipping_rates_response( $package );
+					$services 			  = array(
+						array(
+							'id'               => $service_schema->id,
+							'instance'         => $method->get_instance_id(),
+							'service_settings' => $service_settings,
+						),
+					);
+
+					if ( ! isset( $cached_response_body->rates ) ) {
+						continue;
+					}
+					$tracking_request_body[] = array(
+						'rates'    => $cached_response_body->rates,
```

Comment:
The only part of rates used by the connect server to track the event is rates->id. I'm pretty sure this is the same as $service_schema->id set above. We could eliminate even more code if we only send the services. Then update the code at
https://github.com/Automattic/woocommerce-connect-server/pull/1697/files#diff-112c1dace901da978e2fca31f58eee61289c177ac513d4222a46685261f9fdb4R45
to just look at the service.id instead of rate.id.
Then we don't need the new method get_cached_shipping_rates_response

Reply:
The reason why I decided to send the rates response was because in some cases we make a rates request for a shipping method, but return no rates for that method. For example, when I was testing this, I had FedEx and USPS rates enabled for my shipping zone, but I only received rates for USPS. I counted this as only having USPS live rates involved in the checkout and excluded FedEx, because this seemed like it would make more sense from a store owner/customer's perspective. 

However, that's just me making more assumptions at my own peril. @c-shultz, when we are checking whether a carrier's live rates are involved in a checkout, should we only count the carrier rates that a customer sees at checkout or do we count the carriers that we make requests for? To put that more simply, should we still count a carrier as being involved in a checkout if we don't return any rates for it? 

If the answer to that is a resounding `YES`, then I will be able to shed some more code here and probably put the whole transient trauma to bed once and for all as well.

Reply:
Even if we want to ensure rates for WCS are present we should be able to use `WC()->shipping()->calculate_shipping_for_package()` to get the rates and then extract the service id from the rate id. Calculating the rates again should use the session cache so it won't be very intense. Then we don't need the method `get_cached_shipping_rates_response`.

Reply:
OK, cool. @superdav42 I've now just ripped out all the transient cache commotion, replaced those parts with the `WC()->shipping()->calculate_shipping_for_package()` core function, and changed the endpoint so that we only need to send services instead of rates. This thins out the code even more and reduces the messy reliance on generating that session cache key, which I'm happy with.